frame/contracts/src/exec.rs

```diff
@@ -595,33 +622,46 @@ where
 		gas_limit: Weight,
 		schedule: &Schedule<T>,
 	) -> Result<(Frame<T>, E, Option<u64>), ExecError> {
-		let (account_id, contract_info, executable, entry_point, account_counter) = match frame_args
-		{
-			FrameArgs::Call { dest, cached_info } => {
-				let contract = if let Some(contract) = cached_info {
-					contract
-				} else {
-					<ContractInfoOf<T>>::get(&dest).ok_or(<Error<T>>::ContractNotFound)?
-				};
-
-				let executable = E::from_storage(contract.code_hash, schedule, gas_meter)?;
-
-				(dest, contract, executable, ExportedFunction::Call, None)
-			},
-			FrameArgs::Instantiate { sender, trie_seed, executable, salt } => {
-				let account_id =
-					<Contracts<T>>::contract_address(&sender, executable.code_hash(), &salt);
-				let trie_id = Storage::<T>::generate_trie_id(&account_id, trie_seed);
-				let contract = Storage::<T>::new_contract(
-					&account_id,
-					trie_id,
-					executable.code_hash().clone(),
-				)?;
-				(account_id, contract, executable, ExportedFunction::Constructor, Some(trie_seed))
-			},
-		};
+		let (account_id, contract_info, executable, delegate_caller, entry_point, account_counter) =
+			match frame_args {
+				FrameArgs::Call { dest, cached_info, delegated_call } => {
+					let contract = if let Some(contract) = cached_info {
+						contract
+					} else {
+						<ContractInfoOf<T>>::get(&dest).ok_or(<Error<T>>::ContractNotFound)?
+					};
+
+					let (executable, delegate_caller) =
+						if let Some(DelegatedCall { executable, caller }) = delegated_call {
+							(executable, Some(caller))
+						} else {
+							(E::from_storage(contract.code_hash, schedule, gas_meter)?, None)
+						};
+
+					(dest, contract, executable, delegate_caller, ExportedFunction::Call, None)
+				},
+				FrameArgs::Instantiate { sender, trie_seed, executable, salt } => {
+					let account_id =
+						<Contracts<T>>::contract_address(&sender, executable.code_hash(), &salt);
+					let trie_id = Storage::<T>::generate_trie_id(&account_id, trie_seed);
+					let contract = Storage::<T>::new_contract(
+						&account_id,
+						trie_id,
+						executable.code_hash().clone(),
+					)?;
+					(
+						account_id,
+						contract,
+						executable,
+						None,
+						ExportedFunction::Constructor,
+						Some(trie_seed),
+					)
+				},
+			};
 
 		let frame = Frame {
```

Comment:
When building the frame we have `allows_reentry: true`, but in some other comment I saw it mentions that this should be set to `false` when we're doing a `delegate_call`. Shouldn't this get updated to `false` if `delegated_caller = Some(...)`?

Reply:
AFAIK we agreed on disabling reentry protection i.e. allowing reentry for delegated calls (because you need to trust the callee anyways).

Reply:
Here is where we make the check: frame/contracts/src/wasm/runtime.rs:761
It should be set to false for delegate_call.
We leave it here to be true because we don't want reentrance attack defence enabled.

Reply:
Can you leave a comment in the code explaining this?

Like have:

```rust
// ... short explanation as to why this is okay wrt to delegate_call ...
let allows_reentry = true;

let frame = Frame {
    // ..
    allows_reentry,
    // ...
};
```

Reply:
This suggestion doesn't make sense to me. New frames have it set to `true` unconditionally. Only when you call something the caller frame might switch it to `false`.  This is why we do it in `Ext::call` but not in `Ext::delegate_call`.